labelbox/data/serialization/ndjson/converter.py

```diff
@@ -42,51 +52,68 @@ def serialize(
+                ScalarMetric,
+                ConfusionMatrixMetric,
+                RelationshipAnnotation,
+            ]] = []
+            # First pass to get all RelatiohnshipAnnotaitons
+            # and update the UUIDs of the source and target annotations
             for annotation in label.annotations:
                 if isinstance(annotation, RelationshipAnnotation):
-                    source_uuid = annotation.value.source._uuid
-                    target_uuid = annotation.value.target._uuid
-
-                    if (len(
-                            used_annotation_uuids.intersection(
-                                {source_uuid, target_uuid})) > 0):
-                        new_source_uuid = uuid.uuid4()
-                        new_target_uuid = uuid.uuid4()
-
-                        annotation_uuid_to_generated_uuid_lookup[
-                            source_uuid] = new_source_uuid
-                        annotation_uuid_to_generated_uuid_lookup[
-                            target_uuid] = new_target_uuid
-                        annotation.value.source._uuid = new_source_uuid
-                        annotation.value.target._uuid = new_target_uuid
-                    else:
-                        annotation_uuid_to_generated_uuid_lookup[
-                            source_uuid] = source_uuid
-                        annotation_uuid_to_generated_uuid_lookup[
-                            target_uuid] = target_uuid
-                    used_annotation_uuids.add(annotation._uuid)
-
+                    annotation = copy.deepcopy(annotation)
+                    new_source_uuid = uuid.uuid4()
+                    new_target_uuid = uuid.uuid4()
+                    relationship_uuids[annotation.value.source._uuid].append(
+                        new_source_uuid)
+                    relationship_uuids[annotation.value.target._uuid].append(
+                        new_target_uuid)
+                    annotation.value.source._uuid = new_source_uuid
+                    annotation.value.target._uuid = new_target_uuid
+                    if annotation._uuid in used_uuids:
+                        annotation._uuid = uuid.uuid4()
+                    used_uuids.add(annotation._uuid)
+                    uuid_safe_annotations.append(annotation)
+            # Second pass to update UUIDs for annotations referenced by RelationshipAnnotations
             for annotation in label.annotations:
                 if (not isinstance(annotation, RelationshipAnnotation) and
                         hasattr(annotation, "_uuid")):
-                    annotation._uuid = annotation_uuid_to_generated_uuid_lookup.get(
-                        annotation._uuid, annotation._uuid)
+                    annotation = copy.deepcopy(annotation)
+                    next_uuids = relationship_uuids[annotation._uuid]
+                    if len(next_uuids) > 0:
+                        annotation._uuid = next_uuids.popleft()
```

Comment:
Just so I understand: `popleft` here returns a source annotation for relationship, right? Why do we seem not to deal with the target ?

Reply:
I see I see this line `relationship_uuids[relationship_annotation.value.target.
                                   _uuid].append(new_target_uuid)`